Issue2884 pid autotuning

SenHuang19 opened the pull request:
Hi @JayHuLBL, Could you please kindly take a look at this pull request. I tried my best to eliminate the issues based on the checklist you shared with me

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/package.order:
```diff
@@ -0,0 +1 @@
+Amigo
```
It should be "AMIGO", not "Amigo".

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PID.mo:
```diff
@@ -0,0 +1,83 @@
+within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.Amigo;
```
The package name should change to "Amigo", not "AMIGO"

JayHuLBL commented on Buildings/Controls/OBC/Utilities/BaseClasses/Validation/NormalizedTimeDelay.mo:
```diff
@@ -0,0 +1,38 @@
+within Buildings.Controls.OBC.Utilities.BaseClasses.Validation;
```
The folder `Buildings/Controls/OBC/Utilities/BaseClasses/Validation` should be removed if it is not needed.

SenHuang19 replied:
This folder is removed.

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:
```diff
@@ -0,0 +1,70 @@
+within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.Amigo;
```
The package name should change to "Amigo", not "AMIGO".

SenHuang19 replied:
AMIGO is an abbreviation, similar to PID. So I changed all the "Amigo" into "AMIGO". Please let me know if it makes sense.

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:
```diff
@@ -0,0 +1,70 @@
+within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.Amigo;
+block PI "Identifies the parameters of a PI controller"
```
"Identify parameters of the PI controller"

SenHuang19 replied:
Eliminated all the third-person singular

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:
```diff
@@ -0,0 +1,70 @@
+within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.Amigo;
+block PI "Identifies the parameters of a PI controller"
+  Buildings.Controls.OBC.CDL.Interfaces.RealInput kp(min=1E-6)
+    "Connector for the signal of the gain of a first order time-delayed model"
```
"Gain of a first order time-delayed model"

SenHuang19 replied:
Shorten all the comments

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:
```diff
@@ -0,0 +1,70 @@
+    annotation (Placement(transformation(extent={{-140,40},{-100,80}}),
+        iconTransformation(extent={{-140,40},{-100,80}})));
+  Buildings.Controls.OBC.CDL.Interfaces.RealInput T(min=0)
+    "Connector for the signal of the time constant of a first order time-delayed model"
```
"Time constant of a first order time-delayed model"

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:
```diff
@@ -0,0 +1,70 @@
+    annotation (Placement(transformation(extent={{-140,-20},{-100,20}}),
+        iconTransformation(extent={{-140,-20},{-100,20}})));
+  Buildings.Controls.OBC.CDL.Interfaces.RealInput L(min=1E-6)
+    "Connector for the signal of the time delay of a first order time-delayed model"
```
"Time delay of a first order time-delayed model"

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:
```diff
@@ -0,0 +1,70 @@
+    annotation (Placement(transformation(extent={{-140,-80},{-100,-40}}),
+        iconTransformation(extent={{-140,-80},{-100,-40}})));
+  Buildings.Controls.OBC.CDL.Interfaces.RealOutput k
+    "Connector for control gain signal"
```
"Gain of the controller"

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:
```diff
@@ -0,0 +1,70 @@
+    "Connector for control gain signal"
+    annotation (Placement(transformation(extent={{100,50},{120,70}})));
+  Buildings.Controls.OBC.CDL.Interfaces.RealOutput Ti
+    "Connector for time constant signal for the integral term"
```
"Time constant of the integral term"

JayHuLBL replied:
No need to use "Connector for ...", or "signal for ...". Please revise the comments of other connectors.

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:
```diff
@@ -0,0 +1,70 @@
+          -12,24}}, color={0,0,127}));
+  connect(gain.kp, kp) annotation (Line(points={{-12,36},{-40,36},{-40,60},{
+          -120,60}}, color={0,0,127}));
+  annotation (defaultComponentName = "pI",
```
May use "PI", rather than "pI".

SenHuang19 replied:
replaced "pI" with "PI" and "pID" with "PID"

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:
```diff
@@ -0,0 +1,70 @@
+    annotation (Placement(transformation(extent={{100,50},{120,70}})));
+  Buildings.Controls.OBC.CDL.Interfaces.RealOutput Ti
+    "Connector for time constant signal for the integral term"
+    annotation (Placement(transformation(extent={{100,-50},{120,-30}})));
```
For some reason, your output connectors are smaller than the input connectors. For example in this instance, it is `extent={{100,-50},{120,-30}}`, which means the dimension is 20 by 20. It should be 40 by 40. 
I see in other classes, the output connector are smaller. Please resize them.

SenHuang19 replied:
Modified the size of the output connectors

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/package.mo:
```diff
@@ -0,0 +1,6 @@
+within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner;
+package Amigo "Package with components related to AMIGO tuner"
+annotation (Documentation(info="<html>
+<p>This package contains the blocks to implement the AMIGO tuner. </p>
```
Better to add more detailed introduction of how the "Amigo" tuner work. Please also add the reference link.

SenHuang19 replied:
Added high-level information

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+  Buildings.Controls.OBC.CDL.Discrete.TriggeredSampler samTd(y_start=Td_start) if with_D
+    "Recording the derivative time"
+    annotation (Placement(transformation(extent={{-40,-60},{-20,-80}})));
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.SystemIdentification.FirstOrderTimedelayed.ControlProcessModel
```
the class name should be "ControlProcess", not "ControlProcessModel". It's redundant to add "Model" as the end.

SenHuang19 replied:
This is a little tricky. This class models the "reduced order model" for the control process not the control process itself.

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+    deaBan=deaBan)
+    "A block to approximate the control process"
+    annotation (Placement(transformation(extent={{-20,40},{-40,60}})));
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.Amigo.PI piParameters
```
instance name `piPar`

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.Amigo.PI piParameters
+    if not with_D "A block to calculates the parameters of a PI controller"
+    annotation (Placement(transformation(extent={{-60,70},{-80,90}})));
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.Amigo.PID pidParameters
```
instance name `pidPar`

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+    "A block to approximate the control process"
+    annotation (Placement(transformation(extent={{-20,40},{-40,60}})));
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.Amigo.PI piParameters
+    if not with_D "A block to calculates the parameters of a PI controller"
```
Simplify the comment to `Parameters of the PI controller`

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+    if not with_D "A block to calculates the parameters of a PI controller"
+    annotation (Placement(transformation(extent={{-60,70},{-80,90}})));
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.Amigo.PID pidParameters
+    if with_D "A block to calculates the parameters of a PID controller"
```
Simplify the comment to `Parameters of the PID controller`

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+         - setPoint, yLow=setPoint + yLow)
+    "A block to process the response from the relay controller"
+    annotation (Placement(transformation(extent={{20,40},{0,60}})));
+  Buildings.Controls.OBC.CDL.Continuous.Sources.ModelTime modTim
```
what if the simulation time starts from negative value, say -5000 seconds?

SenHuang19 replied:
The block calculates the time difference and is not sensitive to the absolute value of the time

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+</li>
+</ul>
+</html>"),
+        defaultComponentName = "pIDWithAutotuningAmigoFOTD",
```
`PIDWitTun`

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/Controller.mo:
```diff
@@ -0,0 +1,107 @@
+  Buildings.Controls.OBC.CDL.Interfaces.RealOutput yErr
+    "Control error signal"
+    annotation (Placement(transformation(extent={{100,-10},{120,10}})));
+  Buildings.Controls.OBC.CDL.Logical.OnOffController greMeaSet(final bandwidth=deaBan*2, pre_y_start=true)
```
when assigning the parameters, change line, add `final`

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/Controller.mo:
```diff
@@ -0,0 +1,107 @@
+  connect(controlError.u2, u_s)
+    annotation (Line(points={{-78,14},{-80,14},{-80,
+          0},{-120,0}}, color={0,0,127}));
+  annotation (defaultComponentName = "controller",
```
you may want to change the default name to `relCon`

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/Controller.mo:
```diff
@@ -0,0 +1,107 @@
+          textColor={0,0,255})}),                                Diagram(
+        coordinateSystem(preserveAspectRatio=false)),
+    Documentation(info="<html>
+<p>This block grenerates a real control output signal, y, and a boolean control switch signal, y<sub>On</sub>, as described below:</p>
```
there is no output y<sub>On</sub>, but `On`. Please be consistent between documentation and the code.

SenHuang19 replied:
Fixed it

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo:
```diff
@@ -0,0 +1,180 @@
+    "The minimum value of tOn and tOff"
+    annotation (Placement(transformation(extent={{-80,20},{-60,40}})));
+  Buildings.Controls.OBC.CDL.Continuous.Greater gretOntOff
+    "Check if tOn*tOff*min(tOn*tOff) is larger than 0"
```
avoid use any parameter names in the comments. If in the future the parameter is changed to different name, then we cannot understand the comment.

SenHuang19 replied:
Fixed it and other similar comments as well

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo:
```diff
@@ -0,0 +1,180 @@
+  Buildings.Controls.OBC.CDL.Continuous.Sources.Constant minLen(final k=0)
+    "Minimum value for the horizon length"
+    annotation (Placement(transformation(extent={{-40,20},{-20,40}})));
+  Buildings.Controls.OBC.CDL.Discrete.TriggeredSampler tOnSample(y_start=Buildings.Controls.OBC.CDL.Constants.eps)
```
`final`

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo:
```diff
@@ -0,0 +1,180 @@
+  Buildings.Controls.OBC.CDL.Continuous.Max maxtOntOff
+    "Find the larger one between tOn and tOff"
+    annotation (Placement(transformation(extent={{-20,60},{0,80}})));
+  Buildings.Controls.OBC.CDL.Continuous.Divide halfPeriodRatioCal
```
`halPerRat`

SenHuang19 replied:
ditto

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo:
```diff
@@ -0,0 +1,180 @@
+  Buildings.Controls.OBC.CDL.Continuous.Multiply mul
+    "Detect if the tOn or tOff changes after both of them are larger than 0"
+    annotation (Placement(transformation(extent={{-40,-40},{-20,-20}})));
+  Buildings.Controls.OBC.CDL.Continuous.Greater gretmaxtOntOff
```
the input 1 of the block is from the output of block `AddtOntOff` (the sum of the `tOn` and `tOff`). Is it your intention? I think it should be the output of `max(tOn, tOff)` if it is to implement `Check if either tOn or tOff is larger than 0`.

SenHuang19 replied:
Yes, it was my intention.  I did so because I want to avoid adding additional block of <code>max(tOn, tOff)</code>. When <code>tOn</code> and <code>tOff</code> are both larger than 0, if <code>sum(tOn, tOff)</code> is larger than 0, then <code>On or tOff is larger than 0</code>. I do agree this implementation is confusing, please let me know your thoughts.

JayHuLBL reviewed (changes requested):
@SenHuang19 Please see the inline comments. Please revise the implementation accordingly and I will then review them after you make the changes.

SenHuang19 commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:
```diff
@@ -0,0 +1,70 @@
+</html>", info="<html>
+<p>This block calculates the control gain and the integral time of a PI model.</p>
+<P>Please refer to the following block for detailed information:</p>
+<P><a href=\"modelica://Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.PIGain\">
```
Didn't update this as explained above

SenHuang19 commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+block PIDWithAutotuningAmigoFOTD
+  "A autotuning PID controller with an Amigo tuner and a first order time delayed system model"
+  parameter Real setPoint(min=1E-6) = 0.8
+    "Setpoint for the tuning process";
```
Sorry for the confusion. It is actually the reference value of the relay controller output. I added an assertion to avoid bad values.

SenHuang19 commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+  Buildings.Controls.OBC.CDL.Interfaces.RealOutput y
+    "Connector for actuator output signal"
+    annotation (Placement(transformation(extent={{100,-20},{140,20}}),iconTransformation(extent={{100,-20},{140,20}})));
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.Controller relay(
```
Shorten all the variable names

SenHuang19 commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+    "Connector for actuator output signal"
+    annotation (Placement(transformation(extent={{100,-20},{140,20}}),iconTransformation(extent={{100,-20},{140,20}})));
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.Controller relay(
+    yHig=yHig,
```
Added final to all the places as suggested

SenHuang19 commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+    "Recording the derivative time"
+    annotation (Placement(transformation(extent={{-40,-60},{-20,-80}})));
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.SystemIdentification.FirstOrderTimedelayed.ControlProcessModel
+    controlProcessModel(
```
ditto

SenHuang19 commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,232 @@
+    annotation (Placement(transformation(extent={{-40,-60},{-20,-80}})));
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.SystemIdentification.FirstOrderTimedelayed.ControlProcessModel
+    controlProcessModel(
+    yHig=yHig - setPoint,
```
Added

SenHuang19 commented:
> @SenHuang19 Please see the inline comments. Please revise the implementation accordingly and I will then review them after you make the changes.

@JayHuLBL Thank you so much for your comments. I tried my best to address them. Please let me know if you have more comments and questions.

JayHuLBL commented on Buildings/Resources/Scripts/Dymola/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/Amigo/Validation/PI.mos:
```diff
@@ -0,0 +1,4 @@
+simulateModel("Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.Validation.PI", method="Cvode", tolerance=1e-06, resultFile="PI");
```
Make sure the .mos files path containing "AMIGO", not "Amigo"

SenHuang19 replied:
fixed it

SenHuang19 commented:
@JayHuLBL Could you please let me know if you have more comments?

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:
```diff
@@ -0,0 +1,71 @@
+within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO;
+block PI "Identify the parameters of a PI controller"
+  Buildings.Controls.OBC.CDL.Interfaces.RealInput kp(min=1E-6)
```
add `final`, 
```
final min=1E-6
```

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PID.mo:
```diff
@@ -0,0 +1,83 @@
+within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO;
+block PID "Identifies the parameters of a PID controller"
+  Buildings.Controls.OBC.CDL.Interfaces.RealInput kp(min=1E-6)
```
add `final`

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PIDDerivativeTime.mo:
```diff
@@ -0,0 +1,87 @@
+﻿within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO;
+block PIDDerivativeTime "Identify the derivative time of a PID controller"
+  Buildings.Controls.OBC.CDL.Interfaces.RealInput T(min=0)
```
add `final`

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PIGain.mo:
```diff
@@ -0,0 +1,139 @@
+﻿within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO;
+block PIGain "Identifies the control gain of a PI controller"
+  Buildings.Controls.OBC.CDL.Interfaces.RealInput kp(min=1E-6)
```
add `final`

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/Validation/PI.mo:
```diff
@@ -0,0 +1,318 @@
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.PI PI
+    "Calculate the parameters for a PI controller"
+    annotation (Placement(transformation(extent={{-10,-10},{10,10}})));
+  Buildings.Controls.OBC.CDL.Continuous.Sources.TimeTable RefeDat(table=[0,1,
```
it's hard to maintain the large table in .mo file. Can you please see the model `Buildings.Experimental.DHC.Loads.BaseClasses.Examples.BaseClasses.BuildingTimeSeries` the instance of 'loa` which also shows the time series data, but saved the file in .mos file?

SenHuang19 replied:
Thanks for the information but are we allowed to use blocks outside of the OBC folder?

JayHuLBL replied:
In the validation package, it is ok to use the outside blocks.

mwetter replied:
The validation packages should also only use CDL; you can move the block outside of the OBC package.
(This would be the first time we don't use CDL inside Buildings.OBC.)
@JayHuLBL : The test
https://github.com/lbl-srg/modelica-buildings/blob/b3bc444b85a131748b3a8f9f9b469e9bb0116ae9/Buildings/Resources/Scripts/travis/Makefile#L26
should fail if non-CDL is used in an validation or an example.

SenHuang19 replied:
Thanks for the explanation

JayHuLBL commented on Buildings/Resources/Scripts/Dymola/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/Validation/PIDGain.mos:
```diff
@@ -0,0 +1,4 @@
+simulateModel("Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.Validation.PIDGain", method="Cvode", tolerance=1e-06, resultFile="PIDGain");
+createPlot(id=1, position={15, 15, 1484, 920}, y={"PIDGain.kp", "PIDGain.T", "PIDGain.L"}, range={0.0, 1.0, 0.0, 2.5}, grid=true, subPlot=101, colors={{28,108,200}, {238,46,47}, {0,140,72}});
```
should not be `y={"PIDGain.kp", "PIDGain.T"...`, but `y={"PIDGai.kp", "PIDGai.T"...`.

JayHuLBL replied:
Please run each of the validation model, by clicking `simulate and plot`. It will show you if the model can run and if the variables set in the .mos script are correct.

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,126 @@
+    experiment(
+      StopTime=10000,
+      Tolerance=1e-06,
+      __Dymola_Algorithm="Cvode"),
```
We should not have the `__Dymola_Algorithm` set here. It should be set in the .mos script.

SenHuang19 replied:
* [ ]  in `Buildings.Controls.OBC.CDL.Continuous.Validation.PID`, it should not have `__Dymola_Algorithm="Cvode"`

I couldn't locate the issue, can you provide more information? 

 * [x]  remove the .mos file `Resources/Scripts/Dymola/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIDWithAutotuningAmigoFOTD.mos` and `Resources/Scripts/Dymola/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIWithAutotuningAmigoFOTD.mos`
* [x]  remove `Resources/Scripts/OpenModelica/compareVars/Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation.PIWithAutotuningAmigoFOTD.mos`
 * [x] remove `Resources/Scripts/OpenModelica/compareVars/Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation.PIDWithAutotuningAmigoFOTD.mos`
 * [x]  in folder `Resources/ReferenceResults/Dymola`, remove the reference results for the model `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation.PIWithAutotuningAmigoFOTD` and `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation.PIDWithAutotuningAmigoFOTD`

JayHuLBL replied:
About the `__Dymola_Algorithm`, please check https://github.com/SenHuang19/modelica-buildings/blob/9fe60db3d72d9603b30e6efa6a50e0104ee1b68b/bin/verifyFiles.py#L27.

SenHuang19 replied:
Thanks a lot! I restored the accidental changes.

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,126 @@
+      Tolerance=1e-06,
+      __Dymola_Algorithm="Cvode"),
+    __Dymola_Commands(
+      file="modelica://Buildings/Resources/Scripts/Dymola/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIWithAutotuningAmigoFOTD.mos" "Simulate and plot"),
```
Check the variables in the .mos script are correct.

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIDWithAutotuningAmigoFOTD.mo:
```diff
@@ -0,0 +1,126 @@
+within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation;
+model PIDWithAutotuningAmigoFOTD "Test model for PIDWithAutotuning"
```
This model cannot run.

JayHuLBL commented on Buildings/Resources/Scripts/Dymola/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/Amigo/Validation/PID.mos:
```diff
@@ -0,0 +1,5 @@
+simulateModel("Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.Validation.PID", method="Cvode", tolerance=1e-06, resultFile="PID");
```
delete the folder `Buildings/Resources/Scripts/Dymola/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/Amigo`.

SenHuang19 replied:
done

JayHuLBL reviewed (changes requested):
@SenHuang19 See the inline comments. 
When run `../bin/runUnitTests.py --validate-html-only`, we can see following:
```
The following malformed html syntax has been found:
[-- ./Controls/OBC/CDL/Types/PIDAutoTuner.mo ]
line 39 column 1 - Warning: discarding unexpected </tr>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/Relay/Controller.mo ]
line 21 column 117 - Warning: replacing unexpected i with </i>
line 21 column 129 - Warning: inserting implicit <sub>
line 21 column 144 - Warning: replacing unexpected i with </i>
line 21 column 156 - Warning: inserting implicit <sub>
line 25 column 90 - Warning: replacing unexpected i with </i>
line 25 column 102 - Warning: inserting implicit <sub>
line 21 column 129 - Warning: trimming empty <sub>
line 21 column 156 - Warning: trimming empty <sub>
line 25 column 102 - Warning: trimming empty <sub>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo ]
line 40 column 14 - Warning: replacing unexpected i with </i>
line 40 column 25 - Warning: inserting implicit <sub>
line 40 column 40 - Warning: replacing unexpected i with </i>
line 40 column 51 - Warning: inserting implicit <sub>
line 40 column 25 - Warning: trimming empty <sub>
line 40 column 51 - Warning: trimming empty <sub>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/SystemIdentification/FirstOrderTimedelayed/TimeConstantDelay.mo ]
line 34 column 50 - Warning: trimming empty <p>
line 52 column 47 - Warning: trimming empty <p>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/SystemIdentification/FirstOrderTimedelayed/Gain.mo ]
line 40 column 14 - Warning: replacing unexpected i with </i>
line 40 column 24 - Warning: inserting implicit <sub>
line 40 column 39 - Warning: replacing unexpected i with </i>
line 40 column 49 - Warning: inserting implicit <sub>
line 42 column 8 - Warning: replacing unexpected i with </i>
line 42 column 18 - Warning: inserting implicit <sub>
line 48 column 8 - Warning: replacing unexpected i with </i>
line 48 column 18 - Warning: inserting implicit <sub>
line 56 column 14 - Warning: replacing unexpected i with </i>
line 56 column 26 - Warning: inserting implicit <sub>
line 56 column 41 - Warning: replacing unexpected i with </i>
line 56 column 53 - Warning: inserting implicit <sub>
line 58 column 8 - Warning: replacing unexpected i with </i>
line 58 column 20 - Warning: inserting implicit <sub>
line 60 column 8 - Warning: replacing unexpected i with </i>
line 60 column 19 - Warning: inserting implicit <sub>
line 60 column 34 - Warning: replacing unexpected i with </i>
line 60 column 46 - Warning: inserting implicit <sub>
line 40 column 24 - Warning: trimming empty <sub>
line 40 column 49 - Warning: trimming empty <sub>
line 42 column 18 - Warning: trimming empty <sub>
line 48 column 18 - Warning: trimming empty <sub>
line 56 column 26 - Warning: trimming empty <sub>
line 56 column 53 - Warning: trimming empty <sub>
line 58 column 20 - Warning: trimming empty <sub>
line 60 column 19 - Warning: trimming empty <sub>
line 60 column 46 - Warning: trimming empty <sub>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PIDGain.mo ]
line 40 column 14 - Warning: replacing unexpected i with </i>
line 40 column 24 - Warning: inserting implicit <sub>
line 40 column 24 - Warning: trimming empty <sub>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PIGain.mo ]
line 42 column 11 - Warning: replacing unexpected i with </i>
line 42 column 21 - Warning: inserting implicit <sub>
line 42 column 21 - Warning: trimming empty <sub>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/package.mo ]
line 19 column 1 - Warning: inserting implicit <p>
line 19 column 1 - Warning: trimming empty <p>
```

As suggested previously, please run unit test `../bin/runUnitTests.py -s Buildings.Controls.OBC.Utilities` then you will see which model cannot run and which model did not setup the .mos script properly. Also, run `../bin/verifyFiles.py` so it will show which models have formatting issues.

JayHuLBL commented:
@SenHuang19 See below:
-  [ ] run `../bin/verifyFiles.py`, there are following errors:
```
*** Error:  File 'Controls/OBC/CDL/Continuous/Validation/PID.mo' contains invalid string 'Algorithm='.
*** Error:  File 'Controls/OBC/CDL/Continuous/Validation/PID.mo' contains invalid string '__Dymola_Algorithm'.
*** Error:  File 'Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIWithAutotuningAmigoFOTD.mo' contains invalid string 'Algorithm='.
*** Error:  File 'Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIWithAutotuningAmigoFOTD.mo' contains invalid string '__Dymola_Algorithm'.
*** Error:  File 'Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIDWithAutotuningAmigoFOTD.mo' contains invalid string 'Algorithm='.
*** Error:  File 'Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIDWithAutotuningAmigoFOTD.mo' contains invalid string '__Dymola_Algorithm'.
*** Terminating due to found errors in examined files.
```

- [ ] run `../bin/runUnitTests.py --validate-html-only`, following errors:
```
The following malformed html syntax has been found:
[-- ./Controls/OBC/CDL/Types/PIDAutoTuner.mo ]
line 39 column 1 - Warning: discarding unexpected </tr>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/Relay/Controller.mo ]
line 21 column 117 - Warning: replacing unexpected i with </i>
line 21 column 129 - Warning: inserting implicit <sub>
line 21 column 144 - Warning: replacing unexpected i with </i>
line 21 column 156 - Warning: inserting implicit <sub>
line 25 column 90 - Warning: replacing unexpected i with </i>
line 25 column 102 - Warning: inserting implicit <sub>
line 21 column 129 - Warning: trimming empty <sub>
line 21 column 156 - Warning: trimming empty <sub>
line 25 column 102 - Warning: trimming empty <sub>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo ]
line 40 column 14 - Warning: replacing unexpected i with </i>
line 40 column 25 - Warning: inserting implicit <sub>
line 40 column 40 - Warning: replacing unexpected i with </i>
line 40 column 51 - Warning: inserting implicit <sub>
line 40 column 25 - Warning: trimming empty <sub>
line 40 column 51 - Warning: trimming empty <sub>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/SystemIdentification/FirstOrderTimedelayed/TimeConstantDelay.mo ]
line 34 column 50 - Warning: trimming empty <p>
line 52 column 47 - Warning: trimming empty <p>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/SystemIdentification/FirstOrderTimedelayed/Gain.mo ]
line 40 column 14 - Warning: replacing unexpected i with </i>
line 40 column 24 - Warning: inserting implicit <sub>
line 40 column 39 - Warning: replacing unexpected i with </i>
line 40 column 49 - Warning: inserting implicit <sub>
line 42 column 8 - Warning: replacing unexpected i with </i>
line 42 column 18 - Warning: inserting implicit <sub>
line 48 column 8 - Warning: replacing unexpected i with </i>
line 48 column 18 - Warning: inserting implicit <sub>
line 56 column 14 - Warning: replacing unexpected i with </i>
line 56 column 26 - Warning: inserting implicit <sub>
line 56 column 41 - Warning: replacing unexpected i with </i>
line 56 column 53 - Warning: inserting implicit <sub>
line 58 column 8 - Warning: replacing unexpected i with </i>
line 58 column 20 - Warning: inserting implicit <sub>
line 60 column 8 - Warning: replacing unexpected i with </i>
line 60 column 19 - Warning: inserting implicit <sub>
line 60 column 34 - Warning: replacing unexpected i with </i>
line 60 column 46 - Warning: inserting implicit <sub>
line 40 column 24 - Warning: trimming empty <sub>
line 40 column 49 - Warning: trimming empty <sub>
line 42 column 18 - Warning: trimming empty <sub>
line 48 column 18 - Warning: trimming empty <sub>
line 56 column 26 - Warning: trimming empty <sub>
line 56 column 53 - Warning: trimming empty <sub>
line 58 column 20 - Warning: trimming empty <sub>
line 60 column 19 - Warning: trimming empty <sub>
line 60 column 46 - Warning: trimming empty <sub>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PIDGain.mo ]
line 40 column 14 - Warning: replacing unexpected i with </i>
line 40 column 24 - Warning: inserting implicit <sub>
line 40 column 24 - Warning: trimming empty <sub>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PIGain.mo ]
line 42 column 11 - Warning: replacing unexpected i with </i>
line 42 column 21 - Warning: inserting implicit <sub>
line 42 column 21 - Warning: trimming empty <sub>

[-- ./Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/package.mo ]
line 19 column 1 - Warning: inserting implicit <p>
line 19 column 1 - Warning: trimming empty <p>
```

- [ ] Failed to check model `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation.PIDWithAutotuningAmigoFOTD`, with errors:
<img width="907" alt="Screen Shot 2022-09-22 at 2 50 55 PM" src="https://user-images.githubusercontent.com/23491921/191857866-9123c4f4-afcf-4959-a677-afa103395b96.png">

JayHuLBL commented on Buildings/Resources/Scripts/Dymola/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIDWithAutotuningAmigoFOTD.mos:
```diff
@@ -0,0 +1,5 @@
+simulateModel("Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation.PIDWithAutotuningAmigoFOTD", stopTime=10000, method="Cvode", tolerance=1e-06, resultFile="PIDWithAutotuningAmigoFOTD");
```
This .mos file should be removed.

SenHuang19 replied:
done

JayHuLBL commented on Buildings/Resources/Scripts/Dymola/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIWithAutotuningAmigoFOTD.mos:
```diff
@@ -0,0 +1,5 @@
+simulateModel("Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation.PIWithAutotuningAmigoFOTD", stopTime=10000, method="Cvode", tolerance=1e-06, resultFile="PIWithAutotuningAmigoFOTD");
```
This .mos file should be removed.

SenHuang19 replied:
done

JayHuLBL commented:
@SenHuang19 
- [ ] in `Buildings.Controls.OBC.CDL.Continuous.Validation.PID`, it should not have `__Dymola_Algorithm="Cvode"`
- [ ] remove the .mos file `Resources/Scripts/Dymola/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIDWithAutotuningAmigoFOTD.mos` and `Resources/Scripts/Dymola/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIWithAutotuningAmigoFOTD.mos`
- [ ] remove `Resources/Scripts/OpenModelica/compareVars/Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation.PIWithAutotuningAmigoFOTD.mos`
- [ ] remove `Resources/Scripts/OpenModelica/compareVars/Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation.PIDWithAutotuningAmigoFOTD.mos`
- [ ] in folder `Resources/ReferenceResults/Dymola`, remove the reference results for the model `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation.PIWithAutotuningAmigoFOTD` and `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Validation.PIDWithAutotuningAmigoFOTD`

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PID.mo:
```diff
@@ -0,0 +1,83 @@
+</li>
+</ul>
+</html>", info="<html>
+<p>This block calculates the control gain, the integral time, and the derivative time of a PI model.</p>
```
`...of a PI model`, should be `... of a PID model`.

SenHuang19 replied:
good catch! fixed it

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PIDDerivativeTime.mo:
```diff
@@ -0,0 +1,87 @@
+T<sub>d</sub> = 0.5LT/(0.3L+T),
+</p>
+<p>where <i>T</i> is the time constant of the first-order time-delayed model;</p>
+<p><i>L</i> is the time delay of the first-order time-delayed model.</p>
```
No need to have another paragraph.

SenHuang19 replied:
removed the extra <p></p>

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PIDGain.mo:
```diff
@@ -0,0 +1,95 @@
+    annotation (Placement(transformation(extent={{100,-20},{140,20}})));
+  Buildings.Controls.OBC.CDL.Continuous.Divide div1
+    "Calculate the inverse of the input gain"
+    annotation (Placement(transformation(extent={{-38,40},{-18,80}})));
```
`extent={{-38,40},{-18,80}}` to `extent={{-38,40},{-18,60}}`

SenHuang19 replied:
changed it

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PIDIntegralTime.mo:
```diff
@@ -0,0 +1,100 @@
+  Buildings.Controls.OBC.CDL.Continuous.Add add2
+    "Calculate the sum of the output of gai1 and the output of gai2"
+    annotation (Placement(transformation(extent={{-40,-30},{-20,-10}})));
+  Buildings.Controls.OBC.CDL.Continuous.Divide div
```
This block icon in diagram layer is upside down, so it looks like the `u2/u1`. Please flip it back.

SenHuang19 replied:
done

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PIDIntegralTime.mo:
```diff
@@ -0,0 +1,100 @@
+    "Time constant signal for the integral term"
+    annotation (Placement(transformation(extent={{100,-20},{140,20}})));
+  Buildings.Controls.OBC.CDL.Continuous.Add add1
+    "Calculate the sum of the time delay and the output of gai3"
```
I would avoid to cite any instance name in the comment, like in this one, there is `gai3`. It may not be correct if in the future we change the instance name `gai3` to something else.

SenHuang19 replied:
done

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PIDIntegralTime.mo:
```diff
@@ -0,0 +1,100 @@
+<p>This block calculates the integral time of a PID model.</p>
+<h4>Main equations</h4>
+<p align=\"center\" style=\"font-style:italic;\">
+T<sub>i</sub> = (0.4L + 0.8T)/(L + 0.1T)L,
```
It may be clear if change it to `T<sub>i</sub> = L(0.4L + 0.8T)/(L + 0.1T)`

SenHuang19 replied:
done

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo:
```diff
@@ -0,0 +1,182 @@
+  Buildings.Controls.OBC.CDL.Continuous.Min tmin
+    "Minimum value of the length for the On period and the length for the off period "
+    annotation (Placement(transformation(extent={{-80,20},{-60,40}})));
+  Buildings.Controls.OBC.CDL.Continuous.Greater gretOntOff
```
Does it have the same purpose as instance `gretmaxtOntOff`?

SenHuang19 replied:
No, I modified the comment into "check if both the length for the On period and the length for the off period are larger than 0"

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo:
```diff
@@ -0,0 +1,182 @@
+          -16,8},{-16,30},{-10,30},{-10,38},{-2,38}}, color={0,0,127}));
+  connect(gretmaxtOntOff.u1, mul.u2) annotation (Line(points={{-2,0},{-50,0},{
+          -50,-36},{-42,-36}}, color={0,0,127}));
+  connect(tDecrease.u1, tIncrease.u1) annotation (Line(points={{38,-80},{34,-80},
```
There are some connections that connect input to input, like in this case, connecting the input `tDecrease.u2` to `tIncrease.u2`. It should connect from an instance output to input of another instance.

SenHuang19 replied:
fixed it and other connects as well

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo:
```diff
@@ -0,0 +1,182 @@
+<p>where <i>t<sub>on</sub></i> and <i>t<sub>on</sub></i> are the length of the On period and the Off period, respectively.</p>
+<p>During an On period, the relay switch signal becomes True;</p>
+<p>During an Off period,  the relay switch signal becomes False.</p>
+<p>Note that only the first On period and the first Off period are considered.</p>
```
The code implementation implement more than describe here, like how to trigger the tuning start/end. Need a better documentation.

SenHuang19 replied:
I add an <code>algorithm</code> section, please let me know if more explanations are needed.

JayHuLBL commented on Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo:
```diff
@@ -0,0 +1,182 @@
+<p align=\"center\" style=\"font-style:italic;\">
+&rho; = max(t<sub>on</sub>,t<sub>off</sub>)/ min(t<sub>on</sub>,t<sub>off</sub>),
+</p>
+<p>where <i>t<sub>on</sub></i> and <i>t<sub>on</sub></i> are the length of the On period and the Off period, respectively.</p>
```
second `<i>t<sub>on</sub></I>` should be `<i>t<sub>off</sub></I>`.

JayHuLBL reviewed (changes requested):
@SenHuang19 Please also see the inline comments.

SenHuang19 commented:
@JayHuLBL can you please let me know if the issues are fixed or now?

JayHuLBL commented:
@SenHuang19 Please see comments below:
- [ ] add detailed description to the validation model, please see in the model `Buildings.Controls.OBC.Utilities.Validation.PIDWithInputGains` for the reference of how to describe the validation.
- [ ] in `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.FirstOrderAMIGO`, does the model allow the option of `P`, `PD`? If not, then an assertion should be added to generate warning when user choose these two options. And it seems that the `I` is always needed, so there is no need to have the parameter `with_I`.
- [ ] in `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.HalfPeriodRatio`, there are connection like `connect(triEnd, tOffSample.trigger)`, `connect(tOnSample.trigger, triEnd)`. It should be `connect(tChanges.y, triEnd)`, `connect(tChanges.y, tOffSample.trigger)`,`connect(tChanges.y, tOnSample.trigger)`
- [ ] in the documentation of `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.HalfPeriodRatio`, when saying `We then record the value of the minimum value between (ton+toff)*min(ton,toff) `, do you mean `We then record the value of (ton+toff)*min(ton,toff) at the moment when both ton and toff become larger than 0.`? Better improve the documentation.

SenHuang19 commented:
@JayHuLBL 

Thank you so much for your comments. Please take a look at my response below:
 
@SenHuang19 Please see comments below:
> 
> * [x]  add detailed description to the validation model, please see in the model `Buildings.Controls.OBC.Utilities.Validation.PIDWithInputGains` for the reference of how to describe the validation.

Add detailed descriptions in [Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIDWithFirstOrderAMIGO.mo](https://github.com/lbl-srg/modelica-buildings/pull/3089/commits/82b25f404b76a754d5b88965f27ba5b7f6e762a7#diff-49c927038bb45a096334beafe89a68a66c10467c4086c4f6871ee4af750ca0cc)
[Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Validation/PIWithFirstOrderAMIGO.mo](https://github.com/lbl-srg/modelica-buildings/pull/3089/commits/82b25f404b76a754d5b88965f27ba5b7f6e762a7#diff-71bd81bb9ad87081d629887027417c7a92059171413ecacdb9a67a00425f6bf3)

> * [x]  in `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.FirstOrderAMIGO`, does the model allow the option of `P`, `PD`? If not, then an assertion should be added to generate warning when user choose these two options. And it seems that the `I` is always needed, so there is no need to have the parameter `with_I`.

As suggested, remove <code>with_I</code> and added an assertion

> * [x]  in `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.HalfPeriodRatio`, there are connection like `connect(triEnd, tOffSample.trigger)`, `connect(tOnSample.trigger, triEnd)`. It should be `connect(tChanges.y, triEnd)`, `connect(tChanges.y, tOffSample.trigger)`,`connect(tChanges.y, tOnSample.trigger)`

Fixed the connection issues.

> * [x]  in the documentation of `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.HalfPeriodRatio`, when saying `We then record the value of the minimum value between (ton+toff)*min(ton,toff) `, do you mean `We then record the value of (ton+toff)*min(ton,toff) at the moment when both ton and toff become larger than 0.`? Better improve the documentation.

As suggested, I tried to make it clearer.
